### PR TITLE
[DirectionProvider] support undefined direction

### PR DESCRIPTION
**Pull request by yzimet**
When rendering user-generated content, the direction could sometimes be
undefined. For example, a phone number or an empty string would not be strong
RTL/LTR text.

Before this PR, developers would have to check the direction and conditionally
render a DirectionProvider, which can be combersome.

With this PR `direction` can be undefined, in effect making DirectionProvider a
no-op. Decendent components that use `withDirection` would "skip" undefined
directions and use the context from the closest ancestor that has a defined
direction context (to be verified).

Question: I'm not sure if I need to change `withDirection.jsx` too. Will its `context` reference become stale because it's already subscribed? https://github.com/airbnb/react-with-direction/blob/7cde7e3370d0c4f774e101c42aa08e6a52a37219/src/withDirection.jsx#L41

to: @jasonkb @garrettberg @ljharb @majapw

### CODE REVIEW

**Comment by yzimet** [comment]
To illustrate the goal, here is an example:
```jsx
<DirectionProvider direction={RTL}>
  
  <Carousel />
  
  <DirectionProvider direction={undefined}>
    555
    <PhoneIcon />   // <--- should hopefully have RTL context from ancestor?? But what if direction changes from `undefined` to `LTR`, will the new context be used?
  </DirectionProvider>
  
  <DirectionProvider direction={LTR}>
    input text
  </DirectionProvider>

</DirectionProvider>
```

File: src/DirectionProvider.jsx [hunk 1]
```diff
@@ -12,11 +12,12 @@ import { DIRECTIONS, CHANNEL } from './constants';
 
 const propTypes = forbidExtraProps({
   children: PropTypes.node.isRequired,
-  direction: directionPropType.isRequired,
+  direction: directionPropType,
```
Comment (ljharb) [inline src/DirectionProvider.jsx]: i think i'd prefer using `explicitNull` from `airbnb-prop-types`, and leaving the isRequired, since `undefined` is likely a bug.

**Review by ljharb — CHANGES_REQUESTED** [review]

File: src/DirectionProvider.jsx [hunk 2]
```diff
@@ -33,6 +34,9 @@ export default class DirectionProvider extends React.Component {
   }
 
   getChildContext() {
+    if (!this.props.direction) {
```
Comment (garrettberg) [inline src/DirectionProvider.jsx]: I think this might break if `direction` changes from `null` to LTR or RTL, since `withDirection` on children might be subscribed to a direction broadcast further up the chain. Right?

This might be crazy, but we could wrap `DirectionProvider` in `withDirection`, and set the broadcast state to the direction from context if `this.props.direction` is null. Essentially treating `null` as a sort of "inherit".

Comment (garrettberg) [reply src/DirectionProvider.jsx]: > Question: I'm not sure if I need to change withDirection.jsx too. Will its context reference become stale because it's already subscribed?

Just noticed this. I don't think we can update `withDirection` to handle this since context doesn't trigger a render (which is why we're using brcase to begin with). 😢

File: src/DirectionProvider.jsx (hunk 1 above)
Comment (jasonkb) [reply src/DirectionProvider.jsx]: Hm `undefined` isn't necessarily a bug. E.g. if we use this for an input field, it will frequently start out empty.

Comment (ljharb) [reply src/DirectionProvider.jsx]: Sure - but if you intend it, you could choose to explicitly send `null`, indicating that it's not a bug. Whereas if you omitted it (thus, `undefined`), it might indeed be a bug.

File: src/DirectionProvider.jsx (hunk 2 above)
Comment (garrettberg) [reply src/DirectionProvider.jsx]: @yzimet ^

File: src/DirectionProvider.jsx (hunk 1 above)
Comment (jasonkb) [reply src/DirectionProvider.jsx]: Ah I see, good point! I like the outcome that we change getTextDirection to return `null` instead of `undefined`!

File: src/DirectionProvider.jsx (hunk 2 above)
Comment (yzimet) [reply src/DirectionProvider.jsx]: > This might be crazy, but we could wrap `DirectionProvider` in `withDirection`, and set the broadcast state to the direction from context if `this.props.direction` is null. Essentially treating `null` as a sort of "inherit".

@garrettberg I like your idea. I tried to implement this, but there is a collision on the `direction` prop name from the HOC. Want to pair on this later in the week?

Comment (garrettberg) [reply src/DirectionProvider.jsx]: > Want to pair on this later in the week?

@yzimet Absolutely! 🎉

File: src/DirectionProvider.jsx [hunk 3]
```diff
@@ -48,7 +51,7 @@ export default class DirectionProvider extends React.Component {
     const { children, direction, inline } = this.props;
     const Tag = inline ? 'span' : 'div';
     return (
-      <Tag dir={direction}>
+      <Tag dir={direction || undefined}>
```
Comment (ljharb) [inline src/DirectionProvider.jsx]: On html tags, this may or may not be the same; it may need to be `{...(direction && { direction }) }` (test in a browser to make sure)

**Review by ljharb — COMMENTED** [review]
Even though we don’t normally test propTypes, let’s add a test that passing undefined to DirectionProvider throws?

**Comment by yzimet** [comment]
@jasonkb @garrettberg @ljharb @majapw PTAL! 

Garrett and I paired to implement the inherited direction functionality. We used a similar technique as withDirection to subscribe to the direction context from an ancestor DirectionProvider and save it to local state (`inheritedDirection`). Then, we can broadcast the correct direction in any circumstance:
* direction changes from LTR to RTL (as before)
* direction changes from RTL to LTR (as before)
* direction changes from LTR/RTL to null (broadcasts inherited direction)
* direction changes from null to LTR/RTL (broadcasts this new direction prop)
* direction is null and the inherited direction context changes (broadcasts inherited direction)
* direction is null and the inherited direction context is null (broadcasts default direction: LTR)

We verified that setting a `dir={null}` attribute on a div in React will _not_ render that attribute in DOM. @garrettberg , could you paste your screencapture of this?

File: tests/DirectionProvider_test.jsx [hunk 4]
```diff
@@ -52,12 +52,172 @@ describe('<DirectionProvider>', () => {
...
     expect(broadcastSpy).to.have.callCount(0);
   });
+
+  describe('when direction is omitted', () => {
+    it('throws an error', () => {
+      expect(() => {
+        getWrapper();
+      }).to.throw();
+    });
+  });
```
Comment (yzimet) [inline tests/DirectionProvider_test.jsx:68]: @ljharb this test seems to fail, what is the proper way to test this?

File: tests/DirectionProvider_test.jsx [hunk 5]
```diff
@@ -3,47 +3,47 @@ import { expect } from 'chai';
...
 import DirectionProvider from '../src/DirectionProvider';
-import { DIRECTIONS } from '../src/constants';
+import mockBrcast from './mocks/brcast_mock';
 
-describe('<DirectionProvider>', () => {
-  let children;
-  beforeEach(() => {
-    children = <div>Foo</div>;
-  });
+function getWrapper(props, context) {
```
Comment (ljharb) [inline tests/DirectionProvider_test.jsx:10]: It's very very preferred to repeat the jsx in every test, and *not* to create "getWrapper" functions. In addition, this creates a lot of diff churn, making this PR harder to review. Could we revert this change?

File: src/DirectionProvider.jsx [hunk 6]
```diff
@@ -4,15 +4,15 @@
...
 import brcast from 'brcast';
 import brcastShape from './proptypes/brcast';
 import directionPropType from './proptypes/direction';
-import { DIRECTIONS, CHANNEL } from './constants';
+import { DIRECTIONS, CHANNEL, defaultDirection } from './constants';
 
 const propTypes = forbidExtraProps({
   children: PropTypes.node.isRequired,
-  direction: directionPropType.isRequired,
+  direction: or([directionPropType, explicitNull()]).isRequired,
```
Comment (ljharb) [inline src/DirectionProvider.jsx:15]: I suspect we may need `or([directionPropType.isRequired, explicitNull()])` here?

File: tests/DirectionProvider_test.jsx (hunk 5 above)
Comment (garrettberg) [reply tests/DirectionProvider_test.jsx:10]: My bad! And noted for the future.

**Comment by garrettberg** [comment]
After pairing on this yesterday, I thought of another [potential solution](https://github.com/airbnb/react-with-direction/compare/garrett--inherited-direction) for handling the "inherited" direction.

Basically, `withDirection` can take a `direction` prop, and passes that along instead of context when set. By wrapping `DirectionProvider` in the updated `withDirection`, we get the inherited functionality. It's definitely simpler, but a bit less explicit. Thoughts @yzimet @ljharb?

**Comment by ljharb** [comment]
I'm kind of confused; you want DirectionProvider to also be wrapped by withDirection? that seems badly recursive.

**Comment by garrettberg** [comment]
What we’re essentially adding is support for inheriting direction from an ancestor DirectionProvider when direction is set to null.

The approach here does that by duplicating some code from withDirection and explicitly handling the case of a null prop to use the direction from context.

The approach I linked to leverages withDirection for reading from context instead.

In either case, DirectionProvider needs to access the direction context in addition to providing context to its descendants.

I’m fine with either solution, but I’m not sure how wrapping DirectionProvider in withDirection is any more problematic than accessing context direction as we have here.

**Comment by ljharb** [comment]
Per offline discussion, let's build a new `OptionalDirectionProvider` and `withOverrideableDirection`, and that should enable us to minimize repeated code while not impacting the semantics of the existing exports.

**Comment by yzimet** [comment]
Closing in favor of https://github.com/airbnb/react-with-direction/pull/8
